### PR TITLE
Vision Dashboard Table View and initial explanation flyout

### PR DESCRIPTION
**Pull request by jamesbchao**
Implemented the table view, which is a simple Fluent UI table view. Also created framework for obtaining the explanation on-the-fly for a single image by using a base64 encoded image of the explanation, generated and saved in the responsibleai_vision jupyter notebook. The rest of the flyout to be implemented in fixups/other PR's.

Known issues: 
-changing the images to be obtained from mock data caused the thumbnail slider to break for the image list view
-error/success instances detection in flyout not implemented

Screenshot (33)

Screenshot (31)

## Checklist

- [ ] I have added screenshots above for all UI changes.
- [ ] I have added e2e tests for all UI changes.
- [ ] Documentation was updated if it was needed.

### CODE REVIEW

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        startIndex: this.props.pageSize
+      }
+    ];
+
+    const columns: IColumn[] = [
+      {
+        fieldName: "title",
+        isResizable: true,
+        key: "title",
+        maxWidth: 400,
```
Comment (imatiach-msft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx:95]: can the magic numbers be moved into constants?

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+      {
+        fieldName: "trueY",
+        isResizable: true,
+        key: "truey",
+        maxWidth: 400,
+        minWidth: 200,
+        name: localization.InterpretVision.Dashboard.columnThree
+      },
+      {
+        fieldName: "predictedY",
```
Comment (imatiach-msft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx:116]: nit: it might be cleaner to build the fields in a loop to reduce code duplication

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    index: number | undefined,
+    column?: IColumn | undefined
+  ) => {
+    const value =
+      item && column && column.fieldName
+        ? item[column.fieldName as keyof IListItem]
+        : "";
+
+    const image =
+      item && column && column.fieldName === "title"
```
Comment (imatiach-msft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx:177]: move re-used strings like "title" and "number" to constants

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.tsx [hunk 4]
```diff
@@ -46,7 +66,9 @@ export class VisionExplanationDashboard extends React.Component<
 
     this.state = {
       imageDim: 200,
+      pageSize: 10,
```
Comment (imatiach-msft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.tsx:69]: nit: maybe get this value from the available PageSizeOptions here, to keep it in sync with the list of options

File: libs/localization/src/lib/en.json [hunk 5]
```diff
@@ -1296,6 +1296,9 @@
       "columnFour": "Predicted Y",
       "columnFive": "Other metadata",
       "filter": "Filter",
+      "panelExplanation": "Explanation",
+      "panelInformation": "Information",
+      "panelTitle": "Selected Instance",
```
Comment (imatiach-msft) [inline libs/localization/src/lib/en.json:1301]: nit: I think we try to keep things sentence case, so "Selected instance" is preferred over "Selected Instance"

**Review by imatiach-msft — APPROVED** [review]
Approved with some comments

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+export interface IListItem {
+  title: string;
+  subtitle?: string;
+  image: string;
+  trueY?: number;
+  predictedY?: number;
+  index?: number;
+  other?: number;
```
[the remark this answers is not in the document]
Comment (jamesbchao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx:41]: other metadata the user passes in, but this whole interface is being redesigned as I'm transitioning from mock data to the actual backend data

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.styles.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      fontSize: "large",
+      fontWeight: "600"
+    },
+    featureListContainer: {
+      height: 300,
+      overflow: "auto"
+    },
+    iconContainer: {
+      position: "relative",
+      top: "2px"
```
[the remark this answers is not in the document]
Comment (jamesbchao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.styles.ts:32]: I tried using IStackTokens, but couldn't get the desired effect (literally moving the checkmark down instead of just adding padding around it)

[the remark this answers is not in the document]
Comment (jamesbchao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.styles.ts:32]: I only do this because Fluent UI for some reason renders the icons above the text so it looks slightly off

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                            ? classNames.errorIcon
+                            : classNames.successIcon
+                        }
+                      />
+                    </Stack.Item>
+                    <Stack.Item>
+                      <Text variant="large" className={classNames.title}>
+                        {item?.predictedY !== item?.trueY
+                          ? localization.InterpretVision.Dashboard.titleBarError
+                          : localization.InterpretVision.Dashboard
```
[the remark this answers is not in the document]
Comment (jamesbchao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:97]: I figured that since this text (i.e. "Error Instances") appears in both the dashboard (in the title bar for the image list) and the flyout, I could just reuse it. Should I refactor anyways?

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.tsx [hunk 9]
```diff
@@ -233,6 +298,16 @@ export class VisionExplanationDashboard extends React.Component<
     }
   };
 
+  private onPageSizeSelect = (
+    event: React.FormEvent<HTMLDivElement>,
+    item: IDropdownOption | undefined
+  ): void => {
+    this.setState({ pageSize: Number(item?.text) });
+    if (event) {
```
[the remark this answers is not in the document]
Comment (jamesbchao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.tsx:306]: The linter gets mad at me if I don't include "event: React.FormEvent<HTMLDivElement>" in the parameters, and then gets mad at me if I don't use the variable "event," even though I have no use for it. This is my bandage solution for that, but is there a better way to deal with this?

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+  public render(): React.ReactNode {
+    const classNames = flyoutStyles();
+
+    const item = this.props.item;
+
+    const list = [];
+
+    for (let i = 0; i < 30; i++) {
+      list.push({ title: "Feature 1", value: "value" });
```
Comment (RubyZ10) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:51]: what is this? is it fake data?

Comment (imatiach-msft) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:51]: Yes it's fake data until @jameschao hooks up the backend to pass the metadata features for images

Comment (jameschao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:51]: @imatiach-msft - I think you mean @jamesbchao and not me @jameschao. A great name nonetheless!

Comment (imatiach-msft) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:51]: oh sorry! yes I meant @jamesbchao :)

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+import { IListItem } from "./ImageList";
+
+export interface IFlyoutProps {
+  data: IDatasetSummary;
+  isOpen: boolean;
+  item: IListItem | undefined;
+  callback: () => void;
+}
+
+export class IFlyoutState {}
```
Comment (tongyu-microsoft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:31]: this can be removed

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx (hunk 10 above)
Comment (tongyu-microsoft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:51]: Should title and value be localized here? Could you explain why we set the same value for 30 times?

Comment (imatiach-msft) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:51]: Yes it's fake data until @jamesbchao hooks up the backend to pass the metadata features for images

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+const stackTokens = {
+  childrenGap: "l1"
+};
+
+export class Flyout extends React.Component<IFlyoutProps, IFlyoutState> {
+  public constructor(props: IFlyoutProps) {
```
Comment (tongyu-microsoft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:38]: this constructor can be removed

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    super(props);
+    this.state = {};
+  }
+
+  public render(): React.ReactNode {
+    const classNames = flyoutStyles();
+
+    const item = this.props.item;
```
Comment (tongyu-microsoft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:46]: nit:
`const { data, isOpen, item, callback } = this.props`. So we can directly use the constant without this.props. for all code below

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TitleBar.styles.ts [hunk 14]
```diff
@@ -23,15 +29,15 @@ export const titleBarStyles: () => IProcessedStyleSet<ITitleBarStyles> = () => {
       top: "2px"
     },
     successIcon: {
-      color: "#107c10",
+      color: theme.semanticColors.successIcon,
```
Comment (tongyu-microsoft) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TitleBar.styles.ts:32]: thanks for updating it to use semanticColors!

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (tongyu-microsoft) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.tsx:307]: Please try updating line 303 to `_event: React.FormEvent<HTMLDivElement>,`, this can resolve the lint error. And we can remove `if  (event) {}` here

File: .vscode/settings.json [hunk 15]
```diff
@@ -42,7 +42,7 @@
   },
   "[typescriptreact]": {
     "editor.formatOnSave": true,
-    "editor.defaultFormatter": "esbenp.prettier-vscode"
+    "editor.defaultFormatter": "rvest.vs-code-prettier-eslint"
```
Comment (vinuthakaranth) [inline .vscode/settings.json:45]: why is this needed?

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+            </Stack.Item>
+            <Stack.Item>
+              <Image
+                src={`data:image/jpg;base64,${this.props.data.localExplanations[0]}`}
+                width="800px"
+                style={{ position: "relative", right: 85 }}
+              />
+            </Stack.Item>
+            <Stack.Item>
+              <Separator styles={{ root: { width: "100%" } }} />
```
Comment (vinuthakaranth) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:136]: same here

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+  private onRenderCell = (
+    item?: { title: string; value: string } | undefined
+  ) => {
+    return (
+      <Stack.Item>
+        <Stack
+          horizontal
+          tokens={{ childrenGap: "l2" }}
+          verticalAlign="center"
+          style={{ marginBottom: "20px" }}
```
Comment (vinuthakaranth) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:161]: same here

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+                    imageFit={ImageFit.contain}
+                  />
+                  <Text variant="large" className={classNames.label}>
+                    {item?.title}
+                  </Text>
+                </Stack.Item>
+              </Stack>
+            </Stack.Item>
+            <Stack.Item>
+              <Separator styles={{ root: { width: "100%" } }} />
```
Comment (vinuthakaranth) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:121]: can this be moved to styles file?

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+            <Stack.Item>
+              <Text variant="large" className={classNames.title}>
+                {localization.InterpretVision.Dashboard.panelExplanation}
+              </Text>
+            </Stack.Item>
+            <Stack.Item>
+              <Image
+                src={`data:image/jpg;base64,${this.props.data.localExplanations[0]}`}
+                width="800px"
+                style={{ position: "relative", right: 85 }}
```
Comment (vinuthakaranth) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:132]: same here

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/ImageList.tsx [hunk 20]
```diff
@@ -90,24 +99,29 @@ export class ImageList extends React.Component<
...
+    /* Height of each tile is calculated as rowsPerPage * (defaultImageDim + textPadder + textHeight) */
+    ret.height = 750;
 
-    if (itemIndex === 0) {
-      this.setState({
-        columnCount: Math.ceil(surfaceRect.width / this.props.imageDim)
-      });
-    }
+    /* Item count is calculated as rowsPerPage * (rectangle width / (imageDim + padder + padder)) */
+    ret.itemCount = 3 * (visibleRect.width / (this.props.imageDim + 20 + 20));
```
Comment (vinuthakaranth) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/ImageList.tsx:123]: Thanks for adding the comment:)

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.styles.ts [hunk 21]
```diff
@@ -37,18 +43,18 @@ export const visionExplanationDashboardStyles: () => IProcessedStyleSet<IDataset
...
-        borderTop: "1px solid #EDEBE9"
-      },
       mainContainer: {
         alignItems: "center",
         justifyContent: "space-between",
         width: "100%"
       },
+      mainImageContainer: {
+        height: "650px"
+      },
```
Comment (vinuthakaranth) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.styles.ts:57]: Please check on zoom-in/out if this renders as expected. As We are working on Reflow tasks, it is better to have reflow on newly added code.

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+          closeButtonAriaLabel="Close"
+          onDismiss={this.props.callback}
+          isLightDismiss
+          type={PanelType.custom}
+          customWidth="700px"
+          className={classNames.mainContainer}
+        >
+          <Stack tokens={stackTokens}>
+            <Stack.Item>
+              <Separator styles={{ root: { width: "100%" } }} />
```
Comment (vinuthakaranth) [inline libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/Flyout.tsx:68]: same here

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (jamesbchao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.tsx:307]: thanks!

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+      item && column && column.fieldName === ColumnNames[0].fieldName
+        ? item["image" as keyof IListItem]
+        : "";
+
+    if (typeof image === "number") {
+      return <div />;
+    }
+
+    if (index) {
+      index = index + 1;
```
[the remark this answers is not in the document]
Comment (jamesbchao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/Controls/TableList.tsx:182]: was the same as above with the if (event) return, but I used the same trick you told me and it worked

File: .vscode/settings.json (hunk 15 above)
Comment (jamesbchao) [reply .vscode/settings.json:45]: Oh I didn't know that changed stuff for anyone but me. I'll change it back.

File: libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.styles.ts (hunk 21 above)
Comment (jamesbchao) [reply libs/interpret-vision/src/lib/VisionExplanationDashboard/VisionExplanationDashboard.styles.ts:57]: It works for me! But I can look into the reflow thing if needed

File: .vscode/settings.json (hunk 15 above)
Comment (imatiach-msft) [reply .vscode/settings.json:45]: @jamesbchao can you change this back again?  it looks like it is still there in latest commit

**Review by imatiach-msft — APPROVED** [review]
